projects/RabbitMQ.Client/client/impl/MainSession.cs

```diff
@@ -98,35 +91,31 @@ public override bool HandleFrame(in InboundFrame frame)
         ///</remarks>
         public void SetSessionClosing(bool closeServerInitiated)
         {
-            lock (_closingLock)
+            if (!_closing)
             {
-                if (!_closing)
+                lock (this)
```

Comment:
In general it is not a good practice to lock on this since this can lead to deadlocks because your are locking the whole instance on which someone external could also lock on. Either bring the lock instance object back or use a compare exchange spin approach?

Reply:
Sessions are internal objects not exposed in any way to the public. While I agree with the general statement, I think one does not always have to follow it if the lock from outside is not possible.

Reply:
Yeah, I know locking on the instance is not recommended but in this case we can be sure that no one external will lock on it, unless through reflection gymnastics but all bets are off at that point anyway.

Reply:
I'm a fan of least surprises. Even for internal usage. It doesn't necessarily have to be user code. Sometimes over time things slip in. Having a dedicated lock is the safer choice. Yet I'm not hung up on it. Feel free to ignore

Reply:
We can add a new field just for locking. One more object per connection is not going to make much of a difference.

Reply:
Should we do this?